fix(gatsby-source-drupal): limit concurrentFileRequests while downloading remote images

bakeruk opened the pull request:
- Added fix for **gatsby-source-drupal** regarding `createRemoteFileNode` buffer memory exhaustion.
- Added `activity` timers and logs to show steps during a long build process.

## Description

The issue related to this PR is in the "Download all files" logic. The code is written in **parallel** which causes buffer and memory exhaustion issues when `createRemoteFileNode` is called. This is caused by unresolved promises in the call-stack especially when you have substantial pages and media entities.

The use of `for` and `lodash` loops that do not respect resolving promises before continuing on to the next item in the loop is to blame. They currently overwhelm the call-stack with unresolved promises. 

I have simply refactored the code to be **serialised** using the appropriate for-loop functionality and voila!

## Related Issues

#12848 [gatsby-source-drupal] "Failed to process remote content" error on Drupal images

coreyward commented:
If the issue being experienced is related to concurrency it seems the better solution would be to implement a request pool rather than to abandon concurrency entirely. One potential example: https://github.com/rxaviers/async-pool

bakeruk commented:
Possibly.

With the PR code, 3000+ images takes 208 seconds.

As no one knows what server/instance this code will be running on or what the memory allocation wilp be. It will be hard to configure the pool sizes to make sure this issue don't arise again.

bakeruk commented:
For got to say (*getting off a train)...

I'll give concurrency a go with small pool sizes and see if that works 👍🏼

bakeruk commented:
- Added file request concurrency logic to gatsby-source-drupal to speed build times.
- Added changeable `concurrentFileRequests` option to the plugin and updated `README.md` file

coreyward commented:
What’s your experience with the async pool? Better timing than the synchronous implementation? Does it still reliably resolve the failing downloads?

bakeruk commented:
It's really useful! Cheers for bringing it to my attention.

It's very similar to FTP client functionality, which allows 10+ simultaneous transfers at once and continues to the next item once has completed. 

With just the right pool size, it allows just enough promises in the call-stack that it doesn't exhaust the memory but still give you a semi-non blocking runtime.

3000+ images now take ~86 secs to a win if you ask me =)

Will be rolling this async-pool out in other projects of mine.

bakeruk commented:
@KyleAMathews It would be great if this could get merged in before I take a look at #10339

wardpeet commented on packages/gatsby-source-drupal/src/gatsby-node.js:
```diff
@@ -187,48 +206,62 @@ exports.sourceNodes = async (
+        })
+      } catch (err) {
+        // Ignore
+        console.error(err)
```
do we want to show an error? We ignored this before. If we want this I recommend using report.error and remove the comment

bakeruk replied:
It would be nice (if anything was to fail) to have the visibility of what went wrong.

Will change this to `reporter.error`

wardpeet commented on packages/gatsby-source-drupal/src/gatsby-node.js:
```diff
@@ -187,48 +206,62 @@ exports.sourceNodes = async (
 
-  nodes.forEach(n => createNode(n))
+      // If the fileNode exists set the node ID of the local file
+      if (fileNode && typeof fileNode === `object`) {
```
```suggestion
      if (fileNode) {
```
do we need to do an object check?

wardpeet commented on packages/gatsby-source-drupal/src/gatsby-node.js:
```diff
@@ -1,26 +1,41 @@
   apiBase = apiBase || `jsonapi`
 
+  // Default concurrentFileRequests to `20`
+  concurrentFileRequests = concurrentFileRequests || 20
```
could you explain why 20 is the magic number? 😄

bakeruk replied:
It's dependent on your pipeline's or instance's memory allocation.

20 should be sufficient for all cases. This, of course, can be changed if you know you have sufficient memory to handle the number of promises in the call-stack.

Oddly, in RL testing I found the image size would also affect the success of the runtime if the concurrency pool size was too large.

bakeruk commented on packages/gatsby-source-drupal/src/gatsby-node.js:
```diff
@@ -1,26 +1,41 @@
   // Fetch articles.
   // console.time(`fetch Drupal data`)
-  console.log(`Starting to fetch data from Drupal`)
+  console.info(`Starting to fetch data from Drupal`)
```
Wil do

bakeruk commented:
Amended PR accordingly from @wardpeet's comments above.

wardpeet commented:
Updated PR with reporter from node-api instead.

Tested with our drupal example and works fine.

wardpeet commented:
Published in `gatsby-source-drupal@3.1.5` :tada:
